internal/api/auth/ldap.go

```diff
@@ -1,41 +1,199 @@
+	bindDN    string
+	password  string
+	baseDN    string
 }
 
 func NewLDAPAuth(ldapURL string, bindDN string, password string, baseDN string, tlsConfig *tls.Config) *LDAPAuth {
-	conn, err := ldap.DialURL(ldapURL, ldap.DialWithTLSConfig(tlsConfig))
-	if err != nil {
-		log.Error().Err(err).Str("ldapURL", ldapURL).Msg("failed to connect to the LDAP server")
-		return nil
-	}
 	return &LDAPAuth{
-		ldapClient: conn,
-		bindDN:     bindDN,
-		password:   password,
-		baseDN:     baseDN,
+		tlsConfig: tlsConfig,
+		ldapURL:   ldapURL,
+		bindDN:    bindDN,
+		password:  password,
+		baseDN:    baseDN,
+		reqCh:     make(chan authRequest),
 	}
 }
 
+func SetDefaultTimeout(timeout time.Duration) {
+	ldap.DefaultTimeout = timeout //nolint:reassign  // we want to customize the default timeout
+}
+
+func (l *LDAPAuth) StartWorkers(ctx context.Context, maxWorker int) error {
+	if maxWorker <= 0 {
+		return fmt.Errorf("maxWorker must be greater than 0")
+	}
+	for i := 0; i < maxWorker; i++ {
+		go l.spawnWorker(ctx)
+	}
+	return nil
+}
 func (l *LDAPAuth) AuthenticateUser(username string, password string) bool {
-	if err := l.ldapClient.Bind(l.bindDN, l.password); err != nil {
+	req := authRequest{
+		username: username,
+		password: password,
+		authResp: make(chan bool),
+	}
+	l.reqCh <- req
+	return <-req.authResp
+}
+
+func (l *LDAPAuth) spawnWorker(ctx context.Context) {
+	const maxRetry = 1
+	var conn *ldap.Conn
+	var err error
+	tick := time.NewTicker(time.Minute)
+
+	for {
+		select {
+		case req := <-l.reqCh:
+			auth := false
+			retry := 0
+			for retry <= maxRetry+1 {
+				retry++
+				log.Debug().Msgf("worker LDAP authentication attempt number %d", retry)
+				// (re)connect if needed
+				if conn == nil || conn.IsClosing() {
+					log.Debug().Msg("LDAP connection is closed, reconnecting")
+					conn, err = l.connect()
+					if err != nil {
+						log.Error().Err(err).Msg("worker LDAP reconnection failed")
+						req.authResp <- false
+						break
+					}
+				}
+
+				// bind with the user credentials
+				var connState connectionStatus
+				auth, connState = l.authenticateWithTimeout(ctx, conn, req.username, req.password, ldap.DefaultTimeout)
+
+				if connState == connectionClosed {
+					log.Debug().Msg("LDAP connection was closed by the server, closing on client side")
+					if err := conn.Close(); err != nil {
+						log.Error().Err(err).Msg("connection was closed by the server but failed to close on client side")
+					}
+				} else {
+					// LDAP connection is still up, we accept the authentication result
+					log.Debug().Msg("auth response valid")
+					break
+				}
+			}
+
+			log.Debug().Msgf("worker LDAP authentication attempt number %d, result: %t", retry, auth)
+
+			req.authResp <- auth
+			tick.Reset(time.Minute)
+
+		case <-tick.C:
+			// close connection if no request has been made for a minute
+			log.Debug().Msg("timer reached, closing connection")
+			if conn != nil {
+				if err := conn.Close(); err != nil {
```

Comment:
Why don't you use `closeLDAPConnection` here ?

Reply:
at first it was because I wanted to lower the logs as it was too chatty, but it evolved with the implementation
done